tracepoint_decode/src/enumerator.rs

```diff
@@ -0,0 +1,2155 @@
+    tracepoint_name: &'nam str,
+    event_data: &'dat [u8],
+    name_start: u32,
+    name_len: u32,
+    activity_id_start: u32,
+    activity_id_len: u8,
+    header: EventHeader,
+    keyword: u64,
+}
+
+impl<'nam, 'dat> EventHeaderEventInfo<'nam, 'dat> {
+    /// Returns the `tracepoint_name` that was passed to
+    /// `context.enumerate(tracepoint_name, event_data)`, e.g. "ProviderName_L1K2".
+    pub fn tracepoint_name(&self) -> &'nam str {
+        return self.tracepoint_name;
+    }
+
+    /// Returns the `event_data` that was passed to
+    /// `context.enumerate(tracepoint_name, event_data)`.
+    pub fn event_data(&self) -> &'dat [u8] {
+        return self.event_data;
+    }
+
+    /// Returns a formatter for the event's identity, i.e. "ProviderName:EventName".
+    pub fn identity_display(&self) -> IdentityDisplay<'nam, 'dat> {
+        return IdentityDisplay {
+            provider_name: self.provider_name(),
+            name: self.name_bytes(),
+        };
+    }
+
+    /// Returns a formatter for the event's "meta" suffix.
+    ///
+    /// The returned formatter writes event metadata as a comma-separated list of 0 or more
+    /// JSON name-value pairs, e.g. `"level": 5, "keyword": 3` (including the quotation marks).
+    ///
+    /// The included items default to [`PerfMetaOptions::Default`], but can be customized with
+    /// the `meta_options()` property.
+    ///
+    /// One name-value pair is appended for each metadata item that is both requested
+    /// by `meta_options` and has a meaningful value available in the event. For example,
+    /// the "id" metadata item is only appended if the event has a non-zero `Id` value,
+    /// even if the `meta_options` property includes [`PerfMetaOptions::Id`].
+    ///
+    /// The following metadata items are supported:
+    ///
+    /// - `"provider": "MyProviderName"` (off by default)
+    /// - `"event": "MyEventName"` (off by default)
+    /// - `"id": 123` (omitted if zero)
+    /// - `"version": 1` (omitted if zero)
+    /// - `"level": 5` (omitted if zero)
+    /// - `"keyword": "0x1"` (omitted if zero)
+    /// - `"opcode": 1` (omitted if zero)
+    /// - `"tag": "0x123"` (omitted if zero)
+    /// - `"activity": "12345678-1234-1234-1234-1234567890AB"` (omitted if not present)
+    /// - `"relatedActivity": "12345678-1234-1234-1234-1234567890AB"` (omitted if not present)
+    /// - `"options": "Gmygroup"` (omitted if not present, off by default)
+    /// - `"flags": "0x7"` (omitted if zero, off by default)
+    pub fn json_meta_display(&self) -> JsonMetaDisplay {
+        return JsonMetaDisplay {
+            event_info: self,
+            add_comma_before_first_item: false,
+            meta_options: PerfMetaOptions::Default,
+            convert_options: PerfConvertOptions::Default,
+        };
+    }
+
+    /// Returns the offset into `event_data` where the event name starts.
+    pub fn name_start(&self) -> u32 {
+        return self.name_start;
```

Comment:
This is the point that I reached in the PR when I tired of seeing "return" everywhere and must now protest. Rust code should follow Rust coding conventions.

Reply:
I tried it, and it led to some mistakes and some confusing code and some bugs, so I went back to return everywhere. Maybe I'll try it again later.